src/serac/physics/solid_functional.hpp

```diff
@@ -0,0 +1,579 @@
+      is_quasistatic_ = false;
+    } else {
+      is_quasistatic_ = true;
+    }
+
+    int true_size = velocity_.space().TrueVSize();
+
+    u_.SetSize(true_size);
+    du_dt_.SetSize(true_size);
+    previous_.SetSize(true_size);
+    previous_ = 0.0;
+
+    zero_.SetSize(true_size);
+    zero_ = 0.0;
+  }
+
+  /// @brief Destroy the Solid Functional object
+  ~SolidFunctional()
+  {
+    // Update the mesh with the new deformed nodes if requested
+    if (keep_deformation_ == FinalMeshOption::Deformed) {
+      *reference_nodes_ += displacement_.gridFunc();
+    }
+
+    // Build a new grid function to store the mesh nodes post-destruction
+    // NOTE: MFEM will manage the memory of these objects
+
+    auto mesh_fe_coll  = new mfem::H1_FECollection(order_, mesh_.Dimension());
+    auto mesh_fe_space = new mfem::ParFiniteElementSpace(displacement_.space(), &mesh_, mesh_fe_coll);
+    auto mesh_nodes    = new mfem::ParGridFunction(mesh_fe_space);
+    mesh_nodes->MakeOwner(mesh_fe_coll);
+
+    *mesh_nodes = *reference_nodes_;
+
+    // Set the mesh to the newly created nodes object and pass ownership
+    mesh_.NewNodes(*mesh_nodes, true);
+  }
+
+  /**
+   * @brief Set essential displacement boundary conditions (strongly enforced)
+   *
+   * @param[in] disp_bdr The boundary attributes on which to enforce a displacement
+   * @param[in] disp The prescribed boundary displacement function
+   */
+  void setDisplacementBCs(const std::set<int>&                                           disp_bdr,
+                          std::function<void(const mfem::Vector& x, mfem::Vector& disp)> disp)
+  {
+    // Project the coefficient onto the grid function
+    disp_bdr_coef_ = std::make_shared<mfem::VectorFunctionCoefficient>(dim, disp);
+
+    bcs_.addEssential(disp_bdr, disp_bdr_coef_, displacement_);
+  }
+
+  /**
+   * @brief Set the displacement essential boundary conditions on a single component
+   *
+   * @param[in] disp_bdr The set of boundary attributes to set the displacement on
+   * @param[in] disp The vector function containing the set displacement values
+   * @param[in] component The component to set the displacment on
+   */
+  void setDisplacementBCs(const std::set<int>& disp_bdr, std::function<double(const mfem::Vector& x)> disp,
+                          int component)
+  {
+    // Project the coefficient onto the grid function
+    component_disp_bdr_coef_ = std::make_shared<mfem::FunctionCoefficient>(disp);
+
+    bcs_.addEssential(disp_bdr, component_disp_bdr_coef_, displacement_, component);
+  }
+
+  /// @brief Solve the Quasi-static Newton system
+  void quasiStaticSolve() { nonlin_solver_.Mult(zero_, displacement_.trueVec()); }
+
+  /**
+   * @brief Advance the timestep
+   *
+   * @param[inout] dt The timestep to attempt. This will return the actual timestep for adaptive timestepping
+   * schemes
+   * @pre SolidFunctional::completeSetup() must be called prior to this call
```

Comment:
Should we had a boolean to the class to ensure this was called?

Reply:
I think I can do this by just checking if `residual_ == nullptr`. Commit incoming...